### DISCUSSION TITLE
STAT_ANALYSIS_JOB_ARGS settings in StatAnalysis Conf

### DISCUSSION THREAD

**Discussion by BinbinZhou-NOAA** [dtcenter/METplus#1859] [Configuration] [answered]
I tried to use StatAnalysis tool to get monthly average of BSS of GEFS precip forecast from line type PSTD stat files 
I tried following settings in a StatAnalysis conf file : 

STAT_ANALYSIS_JOB_NAME = aggregate 
STAT_ANALYSIS_JOB_ARGS =  -line_type PSTD  -by FCST_VAR,FCST_LEAD,VX_MASK  -out   output.txt

But the error says  PSTD is not in the list of line type for - line_type
In this case,  what is the correct setting for this pair?

Thank!

Binbin Zhou
EMC/NCEP

**Comment by JohnHalleyGotway** [comment]
Hello @BinbinZhou-NOAA, I see you have a question about aggregating PSTD statistics. You tried running `-job aggregate -line_type PSTD` but found that didn't work. And that's correct, the PSTD statistics can't be aggregated directly. Instead, you aggregate the probabilistic contingency tables first, and the derived PSTD stats from the aggregated table.
```
STAT_ANALYSIS_JOB_NAME = aggregate_stat
STAT_ANALYSIS_JOB_ARGS = -line_type PCT -out_line_type PSTD -by FCST_VAR,FCST_LEAD,VX_MASK -out output.txt
```
Does that produce better results?

**Reply by BinbinZhou-NOAA** [reply]
John,

   My stat files contain both PCT and PSTD line type data.  So I can test your suggestion.  
The running is successful  without error,  but BSS value is "NA".  only BSS_SMPL has value 0.3338.  An example of output result is

V10.1.1 GEFS  NA   480000    20221001_120000 20221009_120000 240000   20221001_120000 20221009_120000 APCP_24_ENS_FREQ_gt10 kg/m^2     A24      APCP_24 kg/m^2    A24     CCPA   CONUS   NEAREST     1           ==0.10000   >10        NA         0.05  PSTD       7272       11 0.022965    0.019766   0.026668   0.0023513  0.0098409   0.022437    0.90245 0.014948  0.011472   0.018423       NA          NA          NA   NA   0.3338             0      0.1      0.2      0.3      0.4      0.5      0.6      0.7      0.8       0.9 

I guess the reason is  that the PCT has no climo information.  So only BSS_SMPL has value.  Since the  precip  climo data have been  provided in the grid_stat verification,  its  PSTD record has  both BSS and BSS_SMPL.  We just take BSS instead of BSS_SMPL.   
Is there a way to include climo data in stat_analysis? 

Binbin

**Reply by JohnHalleyGotway** [reply]
@BinbinZhou-NOAA the short answer is no. Here's a GitHub issue from 2016 where we wrote down this detail:
https://github.com/dtcenter/MET/issues/611

But this has never bubbled to the surface since most aggregation/analysis efforts have gone into METviewer instead of Stat-Analysis. I wanted to check to see if/how METviewer is able to aggregate BSS using external climatologies, but don't know where to look for sample data. Can you point me to a database that includes these BSS values so I could look more carefully at it?

A second option, instead of aggregating the underlying PCT contingency tables would be to run a *summary* job for the BSS column. That'd tell you the average BSS value... rather than a true aggregation:
```
STAT_ANALYSIS_JOB_NAME = summary
STAT_ANALYSIS_JOB_ARGS = -line_type PSTD -by FCST_VAR,FCST_LEAD,VX_MASK -column BSS,BSS_SMPL -out output.txt
```

**Reply by BinbinZhou-NOAA** [reply]
I tested with summary but modified -out output.txt to -out_stat  output.stat  since  if using -out option has some error. 
The running is  fine without error,  but the output.txt file has no data inside.  
Checking the log file, there are a bunch of output data. Just copy one line here
 SUMMARY: PSTD      BSS      APCP_24_ENS_FREQ_gt5  960000    CONUS       9  0.94941      0.92867    0.97015         NA       NA  0.026978          NA        NA  0.8952     0.91422    0.94782    0.94947    0.97082    0.97321      0.97649   0.023      0.08129      MEAN  0.94941           0.94941
I think these results should be in the output file. 
What is the problem?

**Comment by JohnHalleyGotway** [comment, marked as answer]
Binbin,

The "summary" job does NOT write to an actual STAT line type. No such line
type exists. That's why the "-out_stat" file is empty.
However, it should be able to write to the "-out" file. So I'd recommend
taking a closer look at the error message you're getting and debugging
further.

John

**Reply by BinbinZhou-NOAA** [reply]
The error message in the log file is
ERROR  : STATAnalysisJob::parse_job_command() -> unrecognized switch "-out" in job command line: -job summary -line_type PSTD -by FCST_VAR,FCST_LEAD,VX_MASK -column BSS,BSS_SMPL -out output.txt

**Reply by georgemccabe** [reply]
Hi @BinbinZhou-NOAA , you cannot set -out as part of the job command. You can set it using STAT_ANALYSIS_OUTPUT_TEMPLATE and optionally STAT_ANALYSIS_OUTPUT_DIR. If you set the template, it will automatically add the -out argument to the stat_analysis command.

**Reply by BinbinZhou-NOAA** [reply]
George,

   Yes, following your suggestion, the output file now can be created. 
Just ran for BSS (-column BSS). Here is header and one line example in the output file
COL_NAME: LINE_TYPE COLUMN FCST_VAR              FCST_LEAD VX_MASK TOTAL    MEAN  MEAN_NCL MEAN_NCU MEAN_BCL MEAN_BCU       STDEV STDEV_BCL STDEV_BCU     MIN     P10     P25     P50     P75     P90        MAX       IQR    RANGE WMO_TYPE WMO_MEAN WMO_WEIGHTED_MEAN 
 SUMMARY: PSTD      BSS    APCP_24_ENS_FREQ_gt10 1080000   CONUS      10 0.97518  0.96161   0.98875       NA       NA  0.018973          NA        NA 0.94226 0.94257 0.96928 0.98396 0.98609 0.99289    0.99392 0.016805   0.05166     MEAN  0.97518           0.97518

I think the column "MEAN" is the averaged value of BSS (0.97518). Please confirm.

Thanks!

**Reply by JohnHalleyGotway** [reply]
Yes, that's correct the mean is average of the BSS values read from the input. Here's a description of the [summary job output](https://met.readthedocs.io/en/latest/Users_Guide/stat-analysis.html?highlight=summary#id8).

**Comment by BinbinZhou-NOAA** [comment]
Thanks!

Binbin

**Reply by JohnHalleyGotway** [reply]
@BinbinZhou-NOAA I'll go ahead and mark this discussion as being answered. Please feel free to add a comment on the MET#611 issue to indicate your interest in getting a better solution for aggregating BSS across multiple cases with Stat-Analysis.
